src/python/pants/core/util_rules/external_tool.py

```diff
@@ -0,0 +1,248 @@
+            f"curl -L $URL | tee >(wc -c) >(shasum -a 256) >/dev/null",
+        )
+        # For backwards compatibility with BinaryToolBase.  Subclasses of BinaryToolBase can
+        # instead subclass this class, and need only switch their --version-digest-mapping to
+        # --known-versions to get rid of the deprecation warning.
+        # Note: After removing this, also remove get_default_versions_and_digests() above.
+        register(
+            "--version-digest-mapping",
+            removal_version="1.30.0.dev0",
+            removal_hint="Use --known-versions instead.",
+            type=dict,
+            default=cls.get_default_versions_and_digests(),
+            fingerprint=True,
+            help="A dict mapping <platform constraint> -> (<version>, <fingerprint>, <size_bytes>)."
+            f'A "platform constraint" is any of {[c.value for c in PlatformConstraint]}, and '
+            "is the platform to fetch the tool for. A platform-independent tool should "
+            f"use {PlatformConstraint.none.value}, while a platform-dependent tool should specify "
+            'all environments it needs to be used for. The "fingerprint" and "size_bytes" '
+            "arguments are the result printed when running `sha256sum` and `wc -c` on "
+            "the downloaded file, respectively.",
+        )
+
+    @classmethod
+    @abstractmethod
+    def generate_url(cls, plat: Platform, version: str) -> str:
+        """Returns the URL for the given version of the tool, runnable on the given os+arch.
+
+        os and arch default to those of the current system.
+
+        Implementations should raise ExternalToolError if they cannot resolve the arguments
+        to a URL.  The raised exception need not have a message - a sensible one will be generated.
+        """
+
+    @classmethod
+    def generate_exe(cls, plat: Platform, version: str) -> str:
+        """Returns the archive path to the given version of the tool.
+
+        If the tool is downloaded directly, not in an archive, this can be left unimplemented.
+        """
+        return ""
+
+    @classmethod
+    def generate_request(
+        cls, plat: Platform, version: str, known_versions: List[str]
+    ) -> ExternalToolRequest:
+        for known_version in known_versions:
+            try:
+                ver, plat_val, sha256, length = (x.strip() for x in known_version.split("|"))
+            except ValueError:
+                raise ExternalToolError(
+                    f"Bad value for --known-versions (see ./pants "
+                    f"help-advanced {cls.options_scope}): {known_version}"
+                )
+            if plat_val == plat.value and ver == version:
+                digest = Digest(fingerprint=sha256, serialized_bytes_length=int(length))
+                try:
+                    url = cls.generate_url(plat, version)
+                    exe = cls.generate_exe(plat, version) or url.rsplit("/", 1)[-1]
```

Comment:
Is `url.rsplit("/", 1)[-1]` a common idiom? I can assume this is essentially doing `os.path.basename()`?

Reply:
`os.path.basename` splits on the path separator. URLs always use `/`, even on windows, so using the path separator is not technically correct.

Reply:
Great! Thanks!

1.